bugsnag-android-core/src/main/java/com/bugsnag/android/Journal.java

```diff
@@ -0,0 +1,170 @@
+package com.bugsnag.android;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.dslplatform.json.DslJson;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A journal represents a series of commands that will manipulate a document.
+ * Sometimes it's more efficient to store a base document and a list of modifications
+ * to me made to it rather than re-serialize the entire document after every modification.
+ */
+public class Journal {
+    @NonNull
+    public final List<Command> commands;
+
+    public Journal() {
+        commands = new LinkedList<>();
+    }
+
+    /**
+     * Apply this journal to a document.
+     *
+     * This method goes through the journal entries one-by-one, applying their modification
+     * commands to the document. This is a destructive operation that will modify the document.
+     *
+     * Note: If a journal command replaces the root object, the resulting document will be a
+     * different object from the one passed in.
+     *
+     * @param document The document to apply to.
+     * @return The finished document (may be a different object from the one passed in).
+     */
+    public @NonNull Object applyTo(@NonNull Object document) {
+        for (Command command : commands) {
+            document = command.apply(document);
+        }
+        return document;
+    }
+
+    /**
+     * Add a journal command to this journal
+     *
+     * @param command The journal command
+     */
+    public void add(@NonNull Command command) {
+        this.commands.add(command);
+    }
+
+    /**
+     * Clear the journal, removing all journal entries.
+     */
+    public void clear() {
+        this.commands.clear();
+    }
+
+    /**
+     * Serialize the journal to a stream.
+     *
+     * @param out The stream to serialize to.
+     * @throws IOException if an IO exception occurs.
+     */
+    public void serialize(@NonNull OutputStream out) throws IOException {
+        for (Command command : commands) {
+            command.serialize(out);
+            out.write(0);
+            out.flush();
```

Comment:
Should this be using a BufferedOutputStream? This would hopefully avoid the need to flush after each command (unless I'm really missing something)

Reply:
We need to flush immediately because this will be written to the shared memory region. The actual serialization of the journal will be done in a worker thread, but the result needs to be deterministic so that I can update the status fields for the native side to know what's going on in case of a crash.